### PR TITLE
Show update to global.json as part of PR description

### PR DESCRIPTION
**Pull request by JohnTortugo**
Relates to: dotnet/arcade#2427

The crux of the change is:

- The addition of a Metadata dictionary to the GitFile class so that we can record interesting information about files that were committed.
- Patch PullRequestActor to use the information below and update the PR description accordingly.

Tested it locally + private repo of mine. Sample PR is here: https://dnceng.visualstudio.com/internal/_git/Cesar2/pullrequest/7118

### CODE REVIEW

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 1]
```diff
@@ -823,7 +825,48 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
...
+
+                List<GitFile> commitedFiles = await darc.CommitUpdatesAsync(targetRepository, newBranchName, coherencyUpdate.deps.Select(du => du.To).ToList(), message.ToString());
+
+                UpdatePRDescriptionDueConfigFiles(commitedFiles, description);
+            }
+        }
+
+        private void UpdatePRDescriptionDueConfigFiles(List<GitFile> commitedFiles, StringBuilder description)
+        {
+            GitFile globalJsonFile = commitedFiles.
```
Comment (mmitche) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:837]: nit: committedFiles

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 2]
```diff
@@ -813,7 +813,9 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
                     dependenciesToCommit.AddRange(coherencyUpdate.deps);
                 }
 
-                await darc.CommitUpdatesAsync(targetRepository, newBranchName, dependenciesToCommit.Select(du => du.To).ToList(), message.ToString());
+                List<GitFile> commitedFiles = await darc.CommitUpdatesAsync(targetRepository, newBranchName, dependenciesToCommit.Select(du => du.To).ToList(), message.ToString());
+
+                UpdatePRDescriptionDueConfigFiles(commitedFiles, description);
```
Comment (mmitche) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:818]: I think it be cleaner to move CalculatePRDescription below the CommitUpdatesAsync and pass in the original update, the coherency update, and the committed files.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 3]
```diff
@@ -823,7 +825,48 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
...
+
+                if (hasSdkVersionUpdate)
+                {
+                    globalJsonSection.AppendLine($"  - Updates the version of .NET SDK used to " +
+                        $"{globalJsonFile.Metadata[GitFileMetadataName.SdkVersionUpdate]}");
+                }
+
+                if (hasToolsDotnetUpdate)
+                {
+                    globalJsonSection.AppendLine($"  - Updates the version of .NET Runtime used to " +
```
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:863]: I'm fuzzy in what the difference is between tools and sdk nodes, but I believe this also points to an SDK and not to a runtime?

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 2 above)
Comment (JohnTortugo) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:818]: Yeah, definitely better. I only realized that after I coded it. I'll modify the code.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 4]
```diff
@@ -823,7 +825,48 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
...
+                List<GitFile> commitedFiles = await darc.CommitUpdatesAsync(targetRepository, newBranchName, coherencyUpdate.deps.Select(du => du.To).ToList(), message.ToString());
+
+                UpdatePRDescriptionDueConfigFiles(commitedFiles, description);
+            }
+        }
+
+        private void UpdatePRDescriptionDueConfigFiles(List<GitFile> commitedFiles, StringBuilder description)
+        {
+            GitFile globalJsonFile = commitedFiles.
+                Where(gf => gf.FilePath.Equals("global.json", StringComparison.OrdinalIgnoreCase)).
```
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:838]: What's the output for an update to the msbuild sdks section? Just an empty header?

Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:838]: Oh nvm, you write the header if there's an SDK update. Maybe we should change it from "Updates to global.json" to Updates to dotnet SDKs used?

Comment (JohnTortugo) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:838]: Yeah, that's it. I believe the updates to msbuild sdks will also show up in the "Updates:" section as a package update?

Yeah, I can definitely change the title.

Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:838]: Yeah, especially in the topic of shortening PR descriptions, I don't think it makes a lot of sense to include it in both places.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 3 above)
Comment (JohnTortugo) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:863]: I checked with Tomas and ChrisC and it's indeed something specific to our infra and it's also the SDK version.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 5]
```diff
@@ -822,11 +822,39 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
...
+
+                if (hasSdkVersionUpdate)
+                {
+                    globalJsonSection.AppendLine($"  - Updates sdk.version to " +
+                        $"{globalJsonFile.Metadata[GitFileMetadataName.SdkVersionUpdate]}");
+                }
+
+                if (hasToolsDotnetUpdate)
+                {
+                    globalJsonSection.AppendLine($"  - Updates the tools.dotnet to " +
```
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:852]:
```suggestion
                    globalJsonSection.AppendLine($"  - Updates tools.dotnet to " +
```

**Comment by riarenas** [comment]
@JohnTortugo Looks like this change is causing the scenario tests to fail with this call stack:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at SubscriptionActorService.PullRequestActorImplementation.UpdatePRDescriptionDueConfigFiles(List`1 commitedFiles, StringBuilder globalJsonSection) in /_/src/Maestro/SubscriptionActorService/PullRequestActor.cs:line 837
   at SubscriptionActorService.PullRequestActorImplementation.CalculatePRDescription(UpdateAssetsParameters update, List`1 deps, List`1 committedFiles, StringBuilder description) in /_/src/Maestro/SubscriptionActorService/PullRequestActor.cs:line 952
   at SubscriptionActorService.PullRequestActorImplementation.CommitUpdatesAsync(List`1 requiredUpdates, StringBuilder description, IRemote darc, String targetRepository, String newBranchName) in /_/src/Maestro/SubscriptionActorService/PullRequestActor.cs:line 817
   at SubscriptionActorService.PullRequestActorImplementation.CreatePullRequestAsync(List`1 updates) in /_/src/Maestro/SubscriptionActorService/PullRequestActor.cs:line 695
   at SubscriptionActorService.PullRequestActorImplementation.CreatePullRequestAsync(List`1 updates) in /_/src/Maestro/SubscriptionActorService/PullRequestActor.cs:line 767
   at SubscriptionActorService.PullRequestActorImplementation.UpdateAssetsAsync(Guid subscriptionId, Int32 buildId, String sourceRepo, String sourceSha, List`1 assets) in /_/src/Maestro/SubscriptionActorService/PullRequestActor.cs:line 606
   at SubscriptionActorService.ActionRunner.InvokeAction[T](IActionTracker target, ActionMethod method, Object[] arguments) in /_/src/Maestro/SubscriptionActorService/ActionRunner/ActionRunner.cs:line 0
```

**Comment by riarenas** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=595187&view=results

**Pull request merged**